cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStoreTest.java

```diff
@@ -0,0 +1,401 @@
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.templates.plugins.PluginClass;
+import co.cask.cdap.api.templates.plugins.PluginPropertyField;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.internal.AppFabricTestHelper;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.test.SlowTests;
+import com.google.common.base.Charsets;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.io.CharStreams;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.BrokenBarrierException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ */
+// suppressing warnings for Bytes.toBytes() when we know the result is not null
+@SuppressWarnings("ConstantConditions")
+public class ArtifactStoreTest {
+  private static ArtifactStore artifactStore;
+
+  @BeforeClass
+  public static void set() throws Exception {
+    artifactStore = AppFabricTestHelper.getInjector().getInstance(ArtifactStore.class);
+  }
+
+  @After
+  public void cleanup() throws IOException {
+    artifactStore.clear(Constants.DEFAULT_NAMESPACE_ID);
+  }
+
+  @Test
+  public void testAddGetSingleArtifact() throws Exception {
+    Id.Artifact artifactId = Id.Artifact.from(Constants.DEFAULT_NAMESPACE_ID, "myplugins", "1.0.0");
+    List<PluginClass> plugins = ImmutableList.of(
+      new PluginClass("atype", "plugin1", "", "c.c.c.plugin1", "cfg", ImmutableMap.<String, PluginPropertyField>of()),
+      new PluginClass("atype", "plugin2", "", "c.c.c.plugin2", "cfg", ImmutableMap.<String, PluginPropertyField>of())
+    );
+    ArtifactMeta artifactMeta = new ArtifactMeta(plugins);
+
+    String artifactContents = "my artifact contents";
+    artifactStore.write(artifactId, artifactMeta, new ByteArrayInputStream(Bytes.toBytes(artifactContents)));
+
+    ArtifactInfo artifactInfo = artifactStore.getArtifact(artifactId);
+    assertEqual(artifactId, artifactMeta, artifactContents, artifactInfo);
+  }
+
+  @Test(expected = ArtifactAlreadyExistsException.class)
+  public void testImmutability() throws Exception {
+    Id.Artifact artifactId = Id.Artifact.from(Constants.DEFAULT_NAMESPACE_ID, "myplugins", "1.0.0");
+    ArtifactMeta artifactMeta = new ArtifactMeta(ImmutableList.<PluginClass>of());
+
+    String artifactContents = "abc123";
+    // this should work
+    artifactStore.write(artifactId, artifactMeta, new ByteArrayInputStream(Bytes.toBytes(artifactContents)));
```

Comment:
that's a good point. I'll wrap in a try-catch it for this first call and fail the test if it catches the exception.